File: src/main/java/beholder_daemon/entity/ContractCode.java
package beholder_daemon.entity;
import jakarta.persistence.*;
import lombok.Data;

import java.io.IOException;
import java.util.Arrays;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

import beholder_daemon.utils.GzipUtils;

@Entity
@Table(name = "contract_code")
public class ContractCode {
    
    @Id
    @Column(name = "address", length = 66, nullable = false)
    private String address;
    
    @Lob
    @JsonIgnore
    @Column(name = "verifiedCode", columnDefinition = "LONGBLOB")
    private byte[] verifiedCode;
    
    @Column(name = "verifiedStatus", length = 50)
    private String verifiedStatus;
    
    @Column(name = "verifiedProxyChain", columnDefinition = "TEXT")
    private String verifiedProxyChain = "";
    
    @Lob
    @JsonIgnore
    @Column(name = "decompiledCode", columnDefinition = "LONGBLOB")
    private byte[] decompiledCode;
    
    @Column(name = "decompiledStatus", length = 50)
    private String decompiledStatus;
    
    @Column(name = "decompiledProxyChain", columnDefinition = "TEXT")
    private String decompiledProxyChain = "";

    @Column(length = 200)
    private String verifiedFilename;

    @Column
    private Short eoa = 0; // 0 contract;1 normal eoa; 2 smart eoa

    @Column(length = 42)
    private String smartEoaAddress;

    @Lob
    @JsonIgnore
    @Column(columnDefinition = "LONGBLOB")
    private byte[] binaryCode;
    
    // Constructors
    public ContractCode() {}
    
    public ContractCode(String address) {
        this.address = address;
    }
    
    // Getters and Setters
    public String getAddress() { return address; }
    public void setAddress(String address) { this.address = address; }
    
    public byte[] getVerifiedCode() { return verifiedCode; }
    public void setVerifiedCode(byte[] verifiedCode) { this.verifiedCode = verifiedCode; }
    
    public String getVerifiedStatus() { return verifiedStatus; }
    public void setVerifiedStatus(String verifiedStatus) { this.verifiedStatus = verifiedStatus; }
    
    public String getVerifiedProxyChain() { return verifiedProxyChain; }
    public void setVerifiedProxyChain(String verifiedProxyChain) { 
        this.verifiedProxyChain = verifiedProxyChain != null ? verifiedProxyChain : "";
    }
    public void setVerifiedFilename(String verifiedFilename) {
        this.verifiedFilename = verifiedFilename;
    }
    public String getVerifiedFilename() {
        return verifiedFilename;
    }
    public byte[] getDecompiledCode() { return decompiledCode; }
    public void setDecompiledCode(byte[] decompiledCode) { this.decompiledCode = decompiledCode; }
    
    public String getDecompiledStatus() { return decompiledStatus; }
    public void setDecompiledStatus(String decompiledStatus) { this.decompiledStatus = decompiledStatus; }
    
    public String getDecompiledProxyChain() { return decompiledProxyChain; }
    public void setDecompiledProxyChain(String decompiledProxyChain) { 
        this.decompiledProxyChain = decompiledProxyChain != null ? decompiledProxyChain : "";
    }
        // 压缩方法
    public void setVerifiedCodeCompressed(String code) throws IOException {
        this.verifiedCode = GzipUtils.compress(code);
    }
    public void setDecompiledCodeCompressed(String code) throws IOException {
        this.decompiledCode = GzipUtils.compress(code);
    }
    public void setBinary(String code) throws IOException {
        this.binaryCode = GzipUtils.compress(code);
    }
    public Short getEoa() { return eoa; }
    public void setEoa(Short eoa) { this.eoa = eoa; }
    // 解压方法
    @JsonProperty("verifiedCode")
    public String getVerifiedCodeDecompressed() throws IOException {
        if (verifiedCode == null) return null;
        return GzipUtils.decompress(verifiedCode);
    }
    @JsonProperty("decompiledCode")
    public String getDecompiledCodeDecompressed() throws IOException {
        if (decompiledCode == null) return null;
        return GzipUtils.decompress(decompiledCode);
    }
    @JsonProperty("binaryCode")
    public String getBinary() throws IOException {
        if (binaryCode == null) return null;
        return GzipUtils.decompress(binaryCode);
    }

    public String getSmartEoaAddress() { return smartEoaAddress; }
    public void setSmartEoaAddress(String address) { this.smartEoaAddress = address; }
}

File: internal/model/contract_code.go
package model

import (
	"bytes"
	"compress/gzip"
	"errors"
	"io"
)

// ContractCode 合约代码实体
type ContractCode struct {
	Address             string `json:"address" gorm:"primary_key"`
	BinaryCode          []byte `json:"-" gorm:"column:binary_code"`
	Binary              string `json:"binary" gorm:"-"`
	Eoa                 int16  `json:"eoa"`
	SmartEoaAddress     string `json:"smartEoaAddress" gorm:"column:smart_eoa_address"`
	VerifiedStatus      string `json:"verifiedStatus" gorm:"column:verified_status"`
	VerifiedCode        []byte `json:"-" gorm:"column:verified_code"`
	VerifiedCodeDecompressed string `json:"verifiedCode" gorm:"-"`
	VerifiedFilename    string `json:"verifiedFilename" gorm:"column:verified_filename"`
	VerifiedProxyChain  string `json:"verifiedProxyChain" gorm:"column:verified_proxy_chain"`
	DecompiledStatus    string `json:"decompiledStatus" gorm:"column:decompiled_status"`
	DecompiledCode      []byte `json:"-" gorm:"column:decompiled_code"`
	DecompiledCodeDecompressed string `json:"decompiledCode" gorm:"-"`
	DecompiledProxyChain string `json:"decompiledProxyChain" gorm:"column:decompiled_proxy_chain"`
}

// TableName 指定表名
func (ContractCode) TableName() string {
	return "contract_code"
}

// AfterFind 查询后解压数据
func (c *ContractCode) AfterFind() error {
	if len(c.BinaryCode) > 0 {
		binary, err := decompress(c.BinaryCode)
		if err == nil {
			c.Binary = binary
		}
	}

	if len(c.VerifiedCode) > 0 {
		verified, err := decompress(c.VerifiedCode)
		if err == nil {
			c.VerifiedCodeDecompressed = verified
		}
	}

	if len(c.DecompiledCode) > 0 {
		decompiled, err := decompress(c.DecompiledCode)
		if err == nil {
			c.DecompiledCodeDecompressed = decompiled
		}
	}

	return nil
}

// BeforeSave 保存前压缩数据
func (c *ContractCode) BeforeSave() error {
	if c.Binary != "" {
		binary, err := compress(c.Binary)
		if err == nil {
			c.BinaryCode = binary
		}
	}

	if c.VerifiedCodeDecompressed != "" {
		verified, err := compress(c.VerifiedCodeDecompressed)
		if err == nil {
			c.VerifiedCode = verified
		}
	}

	if c.DecompiledCodeDecompressed != "" {
		decompiled, err := compress(c.DecompiledCodeDecompressed)
		if err == nil {
			c.DecompiledCode = decompiled
		}
	}

	return nil
}

// compress 压缩字符串
func compress(s string) ([]byte, error) {
	var buf bytes.Buffer
	gw := gzip.NewWriter(&buf)
	if _, err := gw.Write([]byte(s)); err != nil {
		return nil, err
	}
	if err := gw.Close(); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// decompress 解压字节数组
func decompress(data []byte) (string, error) {
	if len(data) == 0 {
		return "", errors.New("empty data")
	}

	gr, err := gzip.NewReader(bytes.NewReader(data))
	if err != nil {
		return "", err
	}
	defer gr.Close()

	s, err := io.ReadAll(gr)
	if err != nil {
		return "", err
	}

	return string(s), nil
}

// SetBinary 设置二进制代码（压缩）
func (c *ContractCode) SetBinary(code string) error {
	binary, err := compress(code)
	if err != nil {
		return err
	}
	c.BinaryCode = binary
	c.Binary = code
	return nil
}

// SetVerifiedCodeCompressed 设置验证代码（压缩）
func (c *ContractCode) SetVerifiedCodeCompressed(code string) error {
	verified, err := compress(code)
	if err != nil {
		return err
	}
	c.VerifiedCode = verified
	c.VerifiedCodeDecompressed = code
	return nil
}

// SetDecompiledCodeCompressed 设置反编译代码（压缩）
func (c *ContractCode) SetDecompiledCodeCompressed(code string) error {
	decompiled, err := compress(code)
	if err != nil {
		return err
	}
	c.DecompiledCode = decompiled
	c.DecompiledCodeDecompressed = code
	return nil
}
